### PR TITLE
CLI improvements

### PR DESCRIPTION
**Pull request by Isaac-DeFrain**
This PR consolidates the `server` and `client` binaries into the main `mina-indexer` binary and

- provides `server` and `client` subcommands
- adds an initial state `Summary` struct
- adds `best_tip` field to `IndexerState` for storing the best block
- expands the available client commands with `best_ledger` and `summary`
- modifies the `best_chain` command to take a `--num` arg
- modifies some log messages/levels
- adds release LTO
- removes server cli flag `--log-stdout` and makes this the default unless `--log-dir` is specified
- adds server cli flag `--db-override`
  - no implementation
  - intention: delete the db on the path provided by `--database-dir` if set, defaults to initializing the indexer state from the db

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1204807559875974
  - https://app.asana.com/0/0/1204758919833965

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -46,7 +40,11 @@ serde_test = "1.0.160"
...
 tracing-appender = "0.2.2"
 tracing-subscriber = "0.3.17"
+chrono = { version = "0.4.26", features = ["serde"] }
 
 [dependencies.tokio]
 version = "1.25.0"
 features = ["full"]
+
+[profile.release]
+lto = true
```
Comment (jenr24G) [inline Cargo.toml:50]: hopefully enabling Link Time Optimizations will increase our performance!

File: src/server/mod.rs [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+    tracing_subscriber::fmt().with_writer(non_blocking).init();
+
+    info!("Initializing indexer state");
+    let mut indexer_state = IndexerState::new(
+        root_hash.clone(),
+        genesis_ledger.ledger,
+        Some(&database_dir),
+    )?;
+
+    // TODO check if db has values and reconstitute state from db block first
```
Comment (jenr24G) [inline src/server/mod.rs:150]: adding to Asana

File: src/server/mod.rs [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+    let mut buffer = Vec::with_capacity(128);
+    let _read = reader.read_until(0, &mut buffer).await?;
+
+    let mut buffers = buffer.split(|byte| *byte == 32);
+    let command = buffers.next().unwrap();
+
+    let command_string = String::from_utf8(command.to_vec())?;
+
+    match command_string.as_str() {
+        "account" => {
```
Comment (jenr24G) [inline src/server/mod.rs:266]: add final `\0` here

**Review by jenr24G — CHANGES_REQUESTED** [review]
Fantastic! Just a few changes, the LGTM!

File: src/server/mod.rs (hunk 2 above)
Comment (jenr24G) [reply src/server/mod.rs:150]: https://app.asana.com/0/1203669416123378/1204808483573283/f

File: src/block/store.rs [hunk 4]
```diff
@@ -35,7 +35,7 @@ impl r2d2::ManageConnection for BlockStore {
 #[derive(Debug)]
 pub struct BlockStoreConn {
     db_path: PathBuf,
-    database: DBWithThreadMode<MultiThreaded>,
+    pub database: DBWithThreadMode<MultiThreaded>,
```
[the remark this answers is not in the document]
Comment (Isaac-DeFrain) [reply src/block/store.rs:38]: Good idea! Will do.

File: src/server/mod.rs (hunk 3 above)
Comment (Isaac-DeFrain) [reply src/server/mod.rs:266]: I got errors with `\0` at the end. I assumed it was because there was a pk before `\0`

File: Cargo.toml [hunk 5]
```diff
@@ -46,7 +40,12 @@ serde_test = "1.0.160"
 tracing = "0.1.37"
 tracing-appender = "0.2.2"
 tracing-subscriber = "0.3.17"
+chrono = { version = "0.4.26", features = ["serde"] }
```
Comment (jenr24G) [inline Cargo.toml]: switch to `time` https://app.asana.com/0/1203669416123378/1204797859320174/f

File: src/server/mod.rs [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+                root_hash,
+                startup_dir,
+                watch_dir,
+                database_dir,
+                log_file,
+            })
+        }
+    }
+}
+
```
Comment (jenr24G) [inline src/server/mod.rs:116]: tracing requires `#[instrument]` attributes

File: src/server/mod.rs [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+            info!("Received best_ledger command");
+            let data_buffer = buffers.next().unwrap();
+            let path = &String::from_utf8(data_buffer[..data_buffer.len() - 1].to_vec())?
+                .parse::<PathBuf>()?;
+            debug!("Writing ledger to {}", path.display());
+            fs::write(path, format!("{ledger:?}")).await?;
+            let bytes = bcs::to_bytes(&format!("Ledger written to {}", path.display()))?;
+            writer.write_all(&bytes).await?;
+        }
+        "summary\0" => {
```
Comment (jenr24G) [inline src/server/mod.rs:309]: why does this one still have a `\0` when the others don't? do we need it or not?

**Review by jenr24G — APPROVED** [review]
I'll approve pending the `time` change, merge whenever you'd like! Also take a look at https://app.asana.com/0/1203669416123378/1204811332601652/f when you get the chance please!

**Comment by Isaac-DeFrain** [comment]
I replace chrono for time

File: src/server/mod.rs (hunk 6 above)
Comment (Isaac-DeFrain) [reply src/server/mod.rs:116]: They barf arg debugs to stdout. We need to find a better way. Maybe just use log?

File: src/server/mod.rs (hunk 7 above)
Comment (Isaac-DeFrain) [reply src/server/mod.rs:309]: It needs it because it doesn't have an y accompanying args

Comment (Isaac-DeFrain) [reply src/server/mod.rs:309]: Without it, client hangs

File: src/server/mod.rs (hunk 6 above)
Comment (jenr24G) [reply src/server/mod.rs:116]: i'll take a look

**Pull request merged**